src/tests/test_pubkey_pqc.h

```diff
@@ -49,22 +30,54 @@ concept PQC_KEM_KAT_Test_Implementation =
  */
-template <detail::PQC_KEM_KAT_Test_Implementation Delegate>
 class PK_PQC_KEM_KAT_Test : public PK_Test {
-   public:
-      PK_PQC_KEM_KAT_Test() : PK_Test(Delegate::algo_name, Delegate::input_file, "Seed,SS,PK,SK,CT") {}
+   protected:
+      /// Type of a KAT vector entry that can be recomputed using the seed
+      enum class VarType { SharedSecret, PublicKey, PrivateKey, Ciphertext };
 
-   private:
-      using Private_Key = typename Delegate::private_key_t;
-      using Public_Key = typename Delegate::public_key_t;
+      PK_PQC_KEM_KAT_Test(const std::string& algo_name, const std::string& input_file) :
+            PK_Test(algo_name, input_file, "Seed,SS,PK,SK,CT") {}
+
+      // --- Callbacks ---
+
+      /// Map a recomputed value to the expected value from the KAT vector (e.g. apply a hash function if Botan's KAT entry is hashed)
+      virtual std::vector<uint8_t> map_value(const std::string& params,
+                                             std::span<const uint8_t> value,
+                                             VarType) const = 0;
+
+      /// Create an RNG that can be used to generate the keypair. @p rng is the DRBG that is used to expand the seed.
+      virtual Fixed_Output_RNG rng_for_keygen(const std::string& params, Botan::RandomNumberGenerator& rng) const = 0;
+
+      /// Create an RNG that can be used to generate the keypair. @p rng is the DRBG that is used to expand the seed.
+      virtual Fixed_Output_RNG rng_for_encapsulation(const std::string& params,
+                                                     Botan::RandomNumberGenerator& rng) const = 0;
+
+      /// Return true if the algorithm with the specified params should be tested
+      virtual bool is_available(const std::string& params) const = 0;
+
+      /// Callback to test the RNG's state after key generation. If not overridden checks that the RNG is empty.
+      virtual void inspect_rng_after_keygen(const std::string& params,
+                                            const Fixed_Output_RNG& rng_keygen,
+                                            Test::Result& result) const {
+         BOTAN_UNUSED(params);
+         result.confirm("All prepared random bits used for key generation", rng_keygen.empty());
+      }
+
+      /// Callback to test the RNG's state after encapsulation. If not overridden checks that the RNG is empty.
+      virtual void inspect_rng_after_encaps(const std::string& params,
+                                            const Fixed_Output_RNG& rng_encaps,
+                                            Test::Result& result) const {
+         BOTAN_UNUSED(params);
+         result.confirm("All prepared random bits used for encapsulation", rng_encaps.empty());
+      }
 
    private:
-      bool skip_this_test(const std::string& header, const VarMap&) override {
+      bool skip_this_test(const std::string& params, const VarMap&) override {
```

Comment:
Due to the (partial) default implementation and the fact that this is delegating to a new interface, I would say this should be marked final:

```suggestion
      bool skip_this_test(const std::string& params, const VarMap&) final {
```